lib/src/default_index_store.rs

```diff
@@ -192,9 +192,17 @@ impl IndexStore for DefaultIndexStore {
 
     fn write_index(
         &self,
-        index: MutableIndex,
+        index: Box<dyn MutableIndex>,
         op_id: &OperationId,
     ) -> Result<Box<dyn ReadonlyIndex>, IndexWriteError> {
+        let index = index
+            .as_any()
+            .downcast_ref::<MutableIndexImpl>()
+            .expect("index to merge in must be a MutableIndexImpl");
+        // TODO: This clone should not be necessary. Maybe we won't need it once trait
+        // upcasting is possible. Then maybe we can upcast Box<dyn MutableIndex> to
+        // Box<dyn Any> and then downcast it to Box<MutableIndexImpl>.
```

Comment:
`fn into_any(self: Box<Self>) -> Box<dyn Any>` can be added to `MutableIndex`. Then, we can downcast it to `Box<MutableIndexImpl>`.
https://doc.rust-lang.org/std/any/trait.Any.html#method.downcast

Reply:
Nice! I had somehow thought that that would make the trait not object safe, but it makes sense that it still is object safe.